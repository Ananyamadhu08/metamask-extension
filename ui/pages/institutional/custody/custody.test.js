import React from 'react';
import PropTypes from 'prop-types';
import sinon from 'sinon';
import { mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import CustodySubview from './custody.component';

// TODO - check why they fail running together
/* eslint-disable mocha/no-skipped-tests */
describe.skip('CustodySubview', function () {
  let wrapper, clock;

  const mockStore = {
    metamask: {
      provider: {
        type: 'test',
      },
    },
    appState: {
      isLoading: false,
    },
  };

  const store = configureMockStore()(mockStore);

  const props = {
    connectCustodyAddresses: sinon.spy(),
    getCustodianAccounts: sinon.stub().callsFake(() => [
      {
        address: '0xAddress',
        name: 'name',
        custodianDetails: {},
        labels: [],
        chainId: 1,
        accountBalance: 1,
      },
    ]),
    getCustodianAccountsByAddress: sinon.spy(),
    getCustodianToken: sinon.stub().callsFake(() => 'testJWT'),
    getCustodianJWTList: async () => ['jwt1'],
    provider: { chainId: 0x1 },
    history: {
      push: sinon.spy(),
    },
    mostRecentOverviewPage: 'test',
    custodians: [
      {
        production: true,
        name: 'name',
        type: 'type',
        iconUrl: 'iconUrl',
        displayName: 'displayName',
      },
    ],
  };

  beforeEach(() => {
    wrapper = mount(
      <Provider store={store}>
        <CustodySubview.WrappedComponent {...props} />
      </Provider>,
      {
        context: {
          t: (str) => str,
          store,
          metricsEvent: () => undefined,
        },
        childContextTypes: {
          t: PropTypes.func,
          store: PropTypes.object,
          metricsEvent: () => undefined,
        },
      },
    );
    clock = sinon.useFakeTimers(1);
  });

  afterEach(() => {
    props.history.push.resetHistory();
    clock.restore();
  });

  it('opens connect custody without any custody selected', () => {
    const selectedCustodyBtn = wrapper.find(
      '[data-testid="custody-connect-button"]',
    );
    expect(Object.keys(selectedCustodyBtn)).toHaveLength(0);
  });

  it('call getCustodianJwtList on custody select and shows account list on connect click', async () => {
    const custodyBtn = wrapper.find('[data-testid="custody-connect-button"]');
    const listPromise = Promise.resolve(['jwt2']);
    const accountsPromise = Promise.resolve([
      {
        address: 'address',
        name: 'name',
        custodianDetails: { walletId: 'walletId' },
        labels: [],
      },
    ]);
    jest
      .spyOn(props, 'getCustodianAccounts')
      .mockImplementation(() => accountsPromise);
    jest
      .spyOn(props, 'getCustodianJWTList')
      .mockImplementation(() => listPromise);

    custodyBtn.first().simulate('click');

    await listPromise;
    wrapper.update();
    await wrapper
      .find('[data-testid="jwt-form-connect-button"]')
      .first()
      .simulate('click');
    await accountsPromise;
    wrapper.update();
    expect(wrapper.find('.custody-account-list')).toBeTruthy();
    wrapper
      .find('[data-testid="custody-account-list-item-radio-button"] input')
      .first()
      .simulate('change', { target: { checked: true } });
    expect(
      wrapper.find(
        '[data-testid="custody-account-list-item-radio-button"] input[checked="true"]',
      ),
    ).toBeTruthy();
    wrapper
      .find('[data-testid="custody-account-connect-button"]')
      .first()
      .simulate('click');
    expect(props.connectCustodyAddresses.called).toBeTruthy();
  });

  it('shows the empty accounts list when all custodian accounts are already added', async () => {
    const newProps = {
      ...props,
      getCustodianAccounts: sinon.stub().callsFake(() => []),
    };
    wrapper = mount(
      <Provider store={store}>
        <CustodySubview.WrappedComponent {...newProps} />
      </Provider>,
      {
        context: {
          t: (str) => str,
          store,
          metricsEvent: () => undefined,
        },
        childContextTypes: {
          t: PropTypes.func,
          store: PropTypes.object,
          metricsEvent: () => undefined,
        },
      },
    );
    const custodyBtn = wrapper.find('[data-testid="custody-connect-button"]');
    const listPromise = Promise.resolve(['jwt2']);
    const accountsPromise = Promise.resolve([]);
    jest
      .spyOn(props, 'getCustodianAccounts')
      .mockImplementation(() => accountsPromise);
    jest
      .spyOn(props, 'getCustodianJWTList')
      .mockImplementation(() => listPromise);
    custodyBtn.first().simulate('click');

    await listPromise;
    wrapper.update();
    await wrapper
      .find('[data-testid="jwt-form-connect-button"]')
      .first()
      .simulate('click');
    await accountsPromise;
    wrapper.update();

    expect(
      wrapper.find('[data-testid="custody-account-list"]').exists(),
    ).toBeFalsy();
    expect(wrapper.find('[data-testid="custody-accounts-empty"]')).toBeTruthy();
  });

  it('shows the accounts list after the jwt token form', async () => {
    const newProps = {
      ...props,
      getCustodianAccountsByAddress: sinon.stub().throwsException('Error 500'),
    };
    wrapper = mount(
      <Provider store={store}>
        <CustodySubview.WrappedComponent {...newProps} />
      </Provider>,
      {
        context: {
          t: (str) => str,
          store,
          metricsEvent: () => undefined,
        },
        childContextTypes: {
          t: PropTypes.func,
          store: PropTypes.object,
          metricsEvent: () => undefined,
        },
      },
    );
    const custodyBtn = wrapper.find('[data-testid="custody-connect-button"]');
    const listPromise = Promise.resolve(['jwt2']);
    const accountsPromise = Promise.resolve([
      {
        address: 'address',
        name: 'name',
        walletId: 'walletId',
        labels: [],
      },
      {
        address: 'address2',
        name: 'name2',
        walletId: 'walletId',
        labels: [],
      },
    ]);
    jest
      .spyOn(props, 'getCustodianAccounts')
      .mockImplementation(() => accountsPromise);
    jest
      .spyOn(props, 'getCustodianJWTList')
      .mockImplementation(() => listPromise);
    custodyBtn.first().simulate('click');

    await listPromise;
    wrapper.update();
    await wrapper
      .find('[data-testid="jwt-form-connect-button"]')
      .first()
      .simulate('click');
    await accountsPromise;
    wrapper.update();
    expect(wrapper.find('[data-testid="custody-account-list"]')).toBeTruthy();
  });

  it('calls getCustodianAccounts on network change', async () => {
    wrapper.find('CustodySubview').props().provider = { chainId: 2 };
    wrapper.update();
    wrapper.find('CustodySubview').instance().handleNetworkChange();
    expect(props.getCustodianAccounts.called).toBeTruthy();
  });

  it('calls getCustodianAccounts while trying to connect when connectRequest is provided', () => {
    const newProps = {
      ...props,
      getCustodianConnectRequest: () => ({
        custodian: 'jupiter',
        token: 'token',
        apiUrl: 'url',
      }),
    };
    wrapper = mount(
      <Provider store={store}>
        <CustodySubview.WrappedComponent {...newProps} />
      </Provider>,
      {
        context: {
          t: (str) => str,
          store,
          metricsEvent: () => undefined,
        },
        childContextTypes: {
          t: PropTypes.func,
          store: PropTypes.object,
          metricsEvent: () => undefined,
        },
      },
    );
    expect(props.getCustodianAccounts.called).toBeTruthy();
  });
});
