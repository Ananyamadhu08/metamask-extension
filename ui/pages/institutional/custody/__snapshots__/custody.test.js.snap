// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustodyPage renders CustodyPage 1`] = `
<div>
  <div
    class="box box--flex-direction-row"
  >
    
    
    <div
      class="box box--sm:padding-7 box--md:padding-2 box--display-flex box--flex-direction-column"
    >
      <button
        aria-label="Back"
        class="box mm-button-icon mm-button-icon--size-sm box--display-flex box--flex-direction-row box--justify-content-center box--align-items-center box--color-icon-default box--background-color-transparent box--rounded-lg"
      >
        <span
          class="box mm-icon mm-icon--size-sm box--display-inline-block box--flex-direction-row box--color-inherit"
          style="mask-image: url('./images/icons/arrow-left.svg');"
        />
      </button>
      <h4
        class="box mm-text mm-text--body-md box--margin-top-4 box--margin-bottom-4 box--flex-direction-row box--color-text-default"
      >
        Custodial Accounts
      </h4>
      <h6
        class="box mm-text mm-text--body-md box--margin-top-2 box--margin-bottom-5 box--flex-direction-row box--color-text-default"
      >
        Please choose the custodian you want to connect in order to add or refresh a token.
      </h6>
      <div
        class="box box--flex-direction-row"
      >
        <ul
          width="full"
        >
          <div
            class="box box--margin-bottom-4 box--padding-3 box--sm:padding-4 box--display-flex box--flex-direction-row box--justify-content-space-between box--align-items-center box--rounded-sm box--border-color-border-default box--border-style-solid box--border-width-1"
          >
            <div
              class="box box--display-flex box--flex-direction-row box--align-items-center"
            >
              <img
                alt="Saturn Custody"
                height="32"
                src="https://saturn-custody-ui.dev.metamask-institutional.io/saturn.svg"
                width="32"
              />
              <p
                class="box mm-text mm-text--body-md box--margin-left-2 box--flex-direction-row box--color-text-default"
              >
                Saturn Custody
              </p>
            </div>
            <button
              class="box mm-text mm-button-base mm-button-base--size-sm mm-button-primary mm-text--body-md box--padding-right-4 box--padding-left-4 box--display-inline-flex box--flex-direction-row box--justify-content-center box--align-items-center box--color-primary-inverse box--background-color-primary-default box--rounded-pill"
              data-testid="custody-connect-button"
            >
              Select
            </button>
          </div>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
